File: Interviews/GFG/random-problems/Medium/Shortest-Distance-in-a-Binary-Maze/Shortest-Distance-in-a-Binary-Maze.py
from typing import List
import queue

class Solution:
    def shortestPath(self, grid: List[List[int]], source: List[int], destination: List[int]) -> int:
        n, m = len(grid), len(grid[0])
        dist = [[1e9 for j in range(m)] for i in range(n)]
        q = queue.Queue()
        q.put(source)
        dist[source[0]][source[1]] = 0
        
        while q.qsize():
            cn = q.get()
            if cn == destination:
                return dist[destination[0]][destination[1]]
            dx, dy = [0, 0, 1, -1], [1, -1, 0, 0]
            for i in range(4):
                nr, nc = cn[0] + dx[i], cn[1] + dy[i]
                if nr < 0 or nr >= n or nc < 0 or nc >= m or grid[nr][nc] != 1:
                    continue
                if dist[nr][nc] > dist[cn[0]][cn[1]] + 1:
                    dist[nr][nc] = dist[cn[0]][cn[1]] + 1;
                    q.put([nr, nc]);
            
        return -1

File: Interviews/GFG/random-problems/Medium/Shortest-Distance-in-a-Binary-Maze/Shortest-Distance-in-a-Binary-Maze.cpp
class Solution {
public:
    int shortestPath(vector<vector<int>> &grid, pair<int, int> source,
                     pair<int, int> destination) {
        int n = grid.size(), m = grid[0].size();
        queue<vector<int>> q;
        q.push({source.first, source.second});
        vector<vector<int>> dist(n, vector<int>(m, 1e9));
        dist[source.first][source.second] = 0;
        while (!q.empty()) {
            vector<int> cn = q.front();
            q.pop();

            int dx[4] = {0, 0, 1, -1}, dy[4] = {1, -1, 0, 0};
            for (int i = 0; i < 4; i++) {
                int nr = cn[0] + dx[i], nc = cn[1] + dy[i];
                if (nr < 0 or nr >= n or nc < 0 or nc >= m or grid[nr][nc] != 1) {
                    continue;
                }
                if (dist[nr][nc] > dist[cn[0]][cn[1]] + 1) {
                    dist[nr][nc] = dist[cn[0]][cn[1]] + 1;
                    q.push({nr, nc});
                }
            }
        }

        if (dist[destination.first][destination.second] != 1e9) {
            return dist[destination.first][destination.second];
        }

        return -1;
    }
};